lib/carbonite/migrations/v4.ex

```diff
@@ -0,0 +1,305 @@
+      THEN
+        RETURN NULL;
+      END IF;
+
+      /* instantiate change row */
+      change_row = ROW(
+        NEXTVAL('#{prefix}.changes_id_seq'),
+        pg_current_xact_id(),
+        LOWER(TG_OP::TEXT),
+        TG_TABLE_SCHEMA::TEXT,
+        TG_TABLE_NAME::TEXT,
+        NULL,
+        NULL,
+        '{}',
+        NULL
+      );
+
+      /* build table_pk */
+      IF trigger_row.primary_key_columns != '{}' THEN
+        IF (TG_OP IN ('INSERT', 'UPDATE')) THEN
+          pk_source := NEW;
+        ELSIF (TG_OP = 'DELETE') THEN
+          pk_source := OLD;
+        END IF;
+
+        change_row.table_pk := '{}';
+
+        FOREACH col_name IN ARRAY trigger_row.primary_key_columns LOOP
+          EXECUTE 'SELECT $1.' || col_name || '::TEXT' USING pk_source INTO pk_col_val;
+          change_row.table_pk := change_row.table_pk || pk_col_val;
+        END LOOP;
+      END IF;
+
+      /* fill in changed data */
+      IF (TG_OP = 'UPDATE') THEN
+        change_row.data = to_jsonb(NEW.*) - trigger_row.excluded_columns;
+
+        FOR old_field IN SELECT * FROM jsonb_each(to_jsonb(OLD.*) - trigger_row.excluded_columns) LOOP
+          IF NOT change_row.data @> jsonb_build_object(old_field.key, old_field.value)
+             THEN change_row.changed := change_row.changed || old_field.key::VARCHAR;
+          END IF;
+        END LOOP;
+
+        IF change_row.changed = '{}' THEN
+          /* All changed fields are ignored. Skip this update. */
+          RETURN NULL;
+        END IF;
+      ELSIF (TG_OP = 'DELETE') THEN
+        change_row.data = to_jsonb(OLD.*) - trigger_row.excluded_columns;
+      ELSIF (TG_OP = 'INSERT') THEN
+        change_row.data = to_jsonb(NEW.*) - trigger_row.excluded_columns;
+      END IF;
+
+      /* filtered columns */
+      FOREACH col_name IN ARRAY trigger_row.filtered_columns LOOP
+        change_row.data = jsonb_set(change_row.data, ('{' || col_name || '}')::TEXT[], jsonb('"[FILTERED]"'));
+      END LOOP;
+
+      /* insert, fail gracefully unless transaction record present or NEXTVAL has never been called */
+      BEGIN
+        change_row.transaction_id = CURRVAL('#{prefix}.transactions_id_seq');
+
+        /* verify that xact_id matches */
+        IF NOT
+          EXISTS(
+            SELECT 1 FROM #{prefix}.transactions
+            WHERE id = change_row.transaction_id AND xact_id = change_row.transaction_xact_id
+          )
+        THEN
+          RAISE USING ERRCODE = 'foreign_key_violation';
+        END IF;
+
+        INSERT INTO #{prefix}.changes VALUES (change_row.*);
+      EXCEPTION WHEN foreign_key_violation OR object_not_in_prerequisite_state THEN
+          RAISE '% on table %.% without prior INSERT into #{prefix}.transactions',
+            TG_OP, TG_TABLE_SCHEMA, TG_TABLE_NAME USING ERRCODE = 'foreign_key_violation';
+      END;
+
+      RETURN NULL;
+    END;
+    $body$
+    LANGUAGE plpgsql;
```

Comment:
I am not very familiar with writing functions/triggers on postgres, but the changes look good to me.